In [129]:
# Import necessary libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [130]:
# Add the parent directory to the system path
import os 
import sys 
sys.path.append(os.path.abspath(os.path.join('..')))

In [131]:
# importing modules from scripts
import scripts.preprocessing as p

In [132]:
# load population data from data
df = pd.read_csv('../data/concat_data.csv')

In [133]:
df.head()

Country Name Country Code  \
0                        Aruba          ABW   
1  Africa Eastern and Southern          AFE   
2                  Afghanistan          AFG   
3   Africa Western and Central          AFW   
4                       Angola          AGO   

                               Indicator Name     Indicator Code       1960  \
0  Population, female (% of total population)  SP.POP.TOTL.FE.ZS  50.858849   
1  Population, female (% of total population)  SP.POP.TOTL.FE.ZS  50.387880   
2  Population, female (% of total population)  SP.POP.TOTL.FE.ZS  48.083054   
3  Population, female (% of total population)  SP.POP.TOTL.FE.ZS  50.179683   
4  Population, female (% of total population)  SP.POP.TOTL.FE.ZS  49.843790   

        1961       1962       1963       1964       1965       1966  \
0  50.850191  50.845065  50.836016  50.826773  50.823381  50.825589   
1  50.388814  50.391276  50.395656  50.400512  50.405510  50.410713   
2  48.165001  48.242372  48.315893  48.385273  48.450285  48.511803   
3  50.194408  50.203157  50.203679  50.204549  50.204173  50.200936   
4  49.700946  49.673797  49.634980  49.576791  49.488336  49.364408   

        1967       1968       1969       1970       1971       1972  \
0  50.826585  50.834861  50.844429  50.862857  50.882413  50.892872   
1  50.413312  50.413630  50.414998  50.423730  50.424335  50.411762   
2  48.570464  48.626005  48.677997  48.726701  48.772846  48.816351   
3  50.203165  50.209982  50.212982  50.206895  50.190393  50.171685   
4  49.213634  49.061033  48.952953  48.979999  49.100274  49.215607   

        1973       1974       1975       1976       1977       1978  \
0  50.920576  50.967882  51.012106  51.059762  51.112015  51.158120   
1  50.395608  50.373201  50.350121  50.335117  50.320733  50.310497   
2  48.856631  48.893531  48.927195  48.958030  48.985824  49.021606   
3  50.154084  50.137214  50.122838  50.111484  50.108052  50.112248   
4  49.325563  49.429489  49.531610  49.633841  49.732098  49.824167   

        1979       1980       1981       1982       1983       1984  \
0  51.203109  51.245443  51.300029  51.316623  51.283702  51.245053   
1  50.301375  50.249665  50.234107  50.292066  50.373230  50.476537   
2  49.078746  49.147412  49.216172  49.304377  49.403368  49.538809   
3  50.117189  50.121167  50.121837  50.121337  50.124192  50.119761   
4  49.909713  49.989454  50.063008  50.130768  50.199861  50.270013   

        1985       1986       1987       1988       1989       1990  \
0  51.192655  51.131628  51.053530  50.966859  50.919562  50.949598   
1  50.594967  50.651383  50.640337  50.645756  50.650867  50.644383   
2  49.710513  49.837794  49.915439  49.964655  49.981496  49.990753   
3  50.109138  50.101000  50.092365  50.078742  50.063588  50.052719   
4  50.334076  50.391713  50.458259  50.529847  50.592299  50.650016   

        1991       1992       1993       1994       1995       1996  \
0  51.068313  51.203841  51.322554  51.448267  51.556132  51.643855   
1  50.636029  50.655412  50.701783  50.741414  50.778877  50.819764   
2  49.996505  50.001845  49.987425  49.963240  49.939028  49.907191   
3  50.046197  50.040800  50.033299  50.025629  50.020582  50.017777   
4  50.692752  50.728672  50.787160  50.851331  50.871566  50.858805   

        1997       1998       1999       2000       2001       2002  \
0  51.718830  51.790647  51.860891  51.928710  52.020597  52.114272   
1  50.833589  50.825554  50.821628  50.808843  50.790463  50.772365   
2  49.873789  49.843320  49.811100  49.775111  49.740205  49.704418   
3  50.017201  50.013269  50.004903  49.996125  49.988018  49.976103   
4  50.846357  50.843965  50.857152  50.867875  50.875299  50.879883   

        2003       2004       2005       2006       2007       2008  \
0  52.182824  52.252512  52.299355  52.314708  52.316943  52.320503   
1  50.752424  50.731596  50.711651  50.690852  50.669994  50.651777   
2  49.669165  49.635085  49.604651  49.579549  49.550982  

### Exploratory Data Analysis

In [134]:
# shape of the data 
df.shape

(2660, 69)

There are 2660 rows and 69 columns in the data.

In [135]:
# Display basic information about the dataset 
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2660 entries, 0 to 2659
Data columns (total 69 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Country Name    2660 non-null   object 
 1   Country Code    2660 non-null   object 
 2   Indicator Name  2660 non-null   object 
 3   Indicator Code  2660 non-null   object 
 4   1960            2525 non-null   float64
 5   1961            2525 non-null   float64
 6   1962            2525 non-null   float64
 7   1963            2525 non-null   float64
 8   1964            2525 non-null   float64
 9   1965            2525 non-null   float64
 10  1966            2525 non-null   float64
 11  1967            2525 non-null   float64
 12  1968            2525 non-null   float64
 13  1969            2525 non-null   float64
 14  1970            2525 non-null   float64
 15  1971            2525 non-null   float64
 16  1972            2525 non-null   float64
 17  1973            2525 non-null   f

There are 65 numerical columns and 4 categorical columns

In [136]:
# Statistical measures for numerical columns
print(df.describe())

               1960          1961          1962          1963          1964  \
count  2.525000e+03  2.525000e+03  2.525000e+03  2.525000e+03  2.525000e+03   
mean   3.638614e+07  3.688780e+07  3.756525e+07  3.840326e+07  3.924833e+07   
std    1.757972e+08  1.777867e+08  1.808257e+08  1.848227e+08  1.888323e+08   
min    0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00  0.000000e+00   
25%    5.056077e+01  5.056572e+01  5.058031e+01  5.057579e+01  5.056749e+01   
50%    4.709300e+04  4.827500e+04  4.950800e+04  5.050100e+04  5.198900e+04   
75%    2.581133e+06  2.609333e+06  2.665887e+06  2.720673e+06  2.762825e+06   
max    3.031517e+09  3.072470e+09  3.126894e+09  3.193470e+09  3.260480e+09   

               1965          1966          1967          1968          1969  \
count  2.525000e+03  2.525000e+03  2.525000e+03  2.525000e+03  2.525000e+03   
mean   4.010573e+07  4.100186e+07  4.189752e+07  4.281844e+07  4.377248e+07   
std    1.929981e+08  1.973987e+08  2.017695e+08  2.

In [137]:
# List of all columns name in the dataset 
print(df.columns)

Index(['Country Name', 'Country Code', 'Indicator Name', 'Indicator Code',
       '1960', '1961', '1962', '1963', '1964', '1965', '1966', '1967', '1968',
       '1969', '1970', '1971', '1972', '1973', '1974', '1975', '1976', '1977',
       '1978', '1979', '1980', '1981', '1982', '1983', '1984', '1985', '1986',
       '1987', '1988', '1989', '1990', '1991', '1992', '1993', '1994', '1995',
       '1996', '1997', '1998', '1999', '2000', '2001', '2002', '2003', '2004',
       '2005', '2006', '2007', '2008', '2009', '2010', '2011', '2012', '2013',
       '2014', '2015', '2016', '2017', '2018', '2019', '2020', '2021', '2022',
       '2023', 'Unnamed: 68'],
      dtype='object')


In [138]:
# Data structure of each columns 
df.dtypes

Country Name       object
Country Code       object
Indicator Name     object
Indicator Code     object
1960              float64
                   ...   
2020              float64
2021              float64
2022              float64
2023              float64
Unnamed: 68       float64
Length: 69, dtype: object

### Data Quality assesment 
The data quality of the dataset is assessed using the following metrics:
- **Completeness**: The dataset is complete as it contains all the required information.
- **Consistency**: The dataset is consistent as it follows a standard format and does not contain any
inconsistencies.

In [139]:
# Check for missing values 
print(df.isnull().sum())

Country Name         0
Country Code         0
Indicator Name       0
Indicator Code       0
1960               135
                  ... 
2020               130
2021               130
2022               130
2023               130
Unnamed: 68       2660
Length: 69, dtype: int64


In [140]:
# Check for duplicates
print(f"Numer of duplicated rows: {df.duplicated().sum()}")

Numer of duplicated rows: 0


In [141]:
# Percentage of missing values 
p.caculate_missing_percentage(df)

The dataset has 6.06% missing values.


The below table presents a summary of missing values in each columns in the dataset.

In [142]:
# Data quality assesment, total missing values and it's percentage
p.check_missing_values(df)

Missing Values  % of Total Values Data Types
Unnamed: 68               2660             100.00    float64
1962                       135               5.08    float64
1963                       135               5.08    float64
1961                       135               5.08    float64
1960                       135               5.08    float64
...                        ...                ...        ...
2023                       130               4.89    float64
Indicator Code               0               0.00     object
Indicator Name               0               0.00     object
Country Code                 0               0.00     object
Country Name                 0               0.00     object

[69 rows x 3 columns]

The dataframes contains 69 columns and 66 columns of them having missing values.

### Handling Missing Values
- Drop column with high percentage missing values > 50%.
  * Unnamed column (100% missing)

In [143]:
# Drop colmuns with high missing values 
df_cleaned = p.drop_high_missing_columns(df)

Dropped columns: ['Unnamed: 68']


This data are scrapped from world bank, so some countries have total or some of missing values in the year column because of no record.

In [144]:
# Display all columns
pd.set_option('display.max_columns', None)

# Identify numerical columns
numerical_columns = df_cleaned.select_dtypes(include=['number']).columns

# Calculate the percentage of missing values for each row, considering only numerical columns
df_cleaned['missing_percentage'] = df_cleaned[numerical_columns].isna().sum(axis=1) * 100 / len(numerical_columns)

# Filter rows with any missing values in numerical columns
filtered_rows = df_cleaned[df_cleaned['missing_percentage'] > 0]

# Display the 'Country Name' and 'missing_percentage' columns
print("\nRows with missing values in numerical columns and their percentages:")
print(filtered_rows[['Country Name', 'missing_percentage']])


Rows with missing values in numerical columns and their percentages:
                  Country Name  missing_percentage
110             Not classified             100.000
376             Not classified             100.000
642             Not classified             100.000
679   St. Martin (French part)             100.000
793                     Kosovo             100.000
...                        ...                 ...
2324        West Bank and Gaza              46.875
2504            Not classified             100.000
2541  St. Martin (French part)             100.000
2590        West Bank and Gaza              46.875
2655                    Kosovo             100.000

[135 rows x 2 columns]


In [145]:
# Drop rows with high missing values
df_cleaned = p.drop_high_missing_rows(df_cleaned)

Dropped Countries: ['Not classified', 'Not classified', 'Not classified', 'St. Martin (French part)', 'Kosovo', 'Not classified', 'St. Martin (French part)', 'Kosovo', 'Not classified', 'Not classified', 'Aruba', 'Africa Eastern and Southern', 'Africa Western and Central', 'Andorra', 'Arab World', 'American Samoa', 'Antigua and Barbuda', 'Bahamas, The', 'Belize', 'Bermuda', 'Barbados', 'Brunei Darussalam', 'Bhutan', 'Botswana', 'Central Europe and the Baltics', 'Channel Islands', 'Comoros', 'Cabo Verde', 'Caribbean small states', 'Curacao', 'Cayman Islands', 'Cyprus', 'Dominica', 'East Asia & Pacific (excluding high income)', 'Early-demographic dividend', 'East Asia & Pacific', 'Europe & Central Asia (excluding high income)', 'Europe & Central Asia', 'Euro area', 'European Union', 'Fragile and conflict affected situations', 'Fiji', 'Faroe Islands', 'Micronesia, Fed. Sts.', 'Gibraltar', 'Grenada', 'Greenland', 'Guam', 'Guyana', 'High income', 'Heavily indebted poor countries (HIPC)', 'I

After cleaning rows with 100 % missing values in numerical columns we have to fill the other with 0.

In [146]:
df_cleaned = df_cleaned.fillna(0)

In [147]:
# Display all columns
pd.set_option('display.max_columns', None)

# Identify numerical columns
numerical_columns = df_cleaned.select_dtypes(include=['number']).columns

# Calculate the percentage of missing values for each row, considering only numerical columns
df_cleaned['missing_percentage'] = df_cleaned[numerical_columns].isna().sum(axis=1) * 100 / len(numerical_columns)

# Filter rows with any missing values in numerical columns
filtered_rows = df_cleaned[df_cleaned['missing_percentage'] > 0]

# Display the 'Country Name' and 'missing_percentage' columns
print("\nRows with missing values in numerical columns and their percentages:")
print(filtered_rows[['Country Name', 'missing_percentage']])


Rows with missing values in numerical columns and their percentages:
Empty DataFrame
Columns: [Country Name, missing_percentage]
Index: []


#### Save Cleaned data to csv file

In [148]:
# Drop the 'missing_percentage' column
df_cleaned = df_cleaned.drop(columns=['missing_percentage'])

# Save cleaned data
df_cleaned.to_csv('../data/population_data.csv')

In [149]:
df_cleaned.head()

Country Name Country Code  \
0                        Aruba          ABW   
1  Africa Eastern and Southern          AFE   
2                  Afghanistan          AFG   
3   Africa Western and Central          AFW   
4                       Angola          AGO   

                               Indicator Name     Indicator Code       1960  \
0  Population, female (% of total population)  SP.POP.TOTL.FE.ZS  50.858849   
1  Population, female (% of total population)  SP.POP.TOTL.FE.ZS  50.387880   
2  Population, female (% of total population)  SP.POP.TOTL.FE.ZS  48.083054   
3  Population, female (% of total population)  SP.POP.TOTL.FE.ZS  50.179683   
4  Population, female (% of total population)  SP.POP.TOTL.FE.ZS  49.843790   

        1961       1962       1963       1964       1965       1966  \
0  50.850191  50.845065  50.836016  50.826773  50.823381  50.825589   
1  50.388814  50.391276  50.395656  50.400512  50.405510  50.410713   
2  48.165001  48.242372  48.315893  48.385273  48.450285  48.511803   
3  50.194408  50.203157  50.203679  50.204549  50.204173  50.200936   
4  49.700946  49.673797  49.634980  49.576791  49.488336  49.364408   

        1967       1968       1969       1970       1971       1972  \
0  50.826585  50.834861  50.844429  50.862857  50.882413  50.892872   
1  50.413312  50.413630  50.414998  50.423730  50.424335  50.411762   
2  48.570464  48.626005  48.677997  48.726701  48.772846  48.816351   
3  50.203165  50.209982  50.212982  50.206895  50.190393  50.171685   
4  49.213634  49.061033  48.952953  48.979999  49.100274  49.215607   

        1973       1974       1975       1976       1977       1978  \
0  50.920576  50.967882  51.012106  51.059762  51.112015  51.158120   
1  50.395608  50.373201  50.350121  50.335117  50.320733  50.310497   
2  48.856631  48.893531  48.927195  48.958030  48.985824  49.021606   
3  50.154084  50.137214  50.122838  50.111484  50.108052  50.112248   
4  49.325563  49.429489  49.531610  49.633841  49.732098  49.824167   

        1979       1980       1981       1982       1983       1984  \
0  51.203109  51.245443  51.300029  51.316623  51.283702  51.245053   
1  50.301375  50.249665  50.234107  50.292066  50.373230  50.476537   
2  49.078746  49.147412  49.216172  49.304377  49.403368  49.538809   
3  50.117189  50.121167  50.121837  50.121337  50.124192  50.119761   
4  49.909713  49.989454  50.063008  50.130768  50.199861  50.270013   

        1985       1986       1987       1988       1989       1990  \
0  51.192655  51.131628  51.053530  50.966859  50.919562  50.949598   
1  50.594967  50.651383  50.640337  50.645756  50.650867  50.644383   
2  49.710513  49.837794  49.915439  49.964655  49.981496  49.990753   
3  50.109138  50.101000  50.092365  50.078742  50.063588  50.052719   
4  50.334076  50.391713  50.458259  50.529847  50.592299  50.650016   

        1991       1992       1993       1994       1995       1996  \
0  51.068313  51.203841  51.322554  51.448267  51.556132  51.643855   
1  50.636029  50.655412  50.701783  50.741414  50.778877  50.819764   
2  49.996505  50.001845  49.987425  49.963240  49.939028  49.907191   
3  50.046197  50.040800  50.033299  50.025629  50.020582  50.017777   
4  50.692752  50.728672  50.787160  50.851331  50.871566  50.858805   

        1997       1998       1999       2000       2001       2002  \
0  51.718830  51.790647  51.860891  51.928710  52.020597  52.114272   
1  50.833589  50.825554  50.821628  50.808843  50.790463  50.772365   
2  49.873789  49.843320  49.811100  49.775111  49.740205  49.704418   
3  50.017201  50.013269  50.004903  49.996125  49.988018  49.976103   
4  50.846357  50.843965  50.857152  50.867875  50.875299  50.879883   

        2003       2004       2005       2006       2007       2008  \
0  52.182824  52.252512  52.299355  52.314708  52.316943  52.320503   
1  50.752424  50.731596  50.711651  50.690852  50.669994  50.651777   
2  49.669165  49.635085  49.604651  49.579549  49.550982  